framework/src/vectorpostprocessors/SamplerBase.C

```diff
@@ -44,104 +44,106 @@ void
 }
 
 void
 SamplerBase::addSample(const Point & p, const Real & id, const std::vector<Real> & values)
 {
-  _x_tmp.push_back(p(0));
-  _y_tmp.push_back(p(1));
-  _z_tmp.push_back(p(2));
+  _x.push_back(p(0));
+  _y.push_back(p(1));
+  _z.push_back(p(2));
 
-  _id_tmp.push_back(id);
+  _id.push_back(id);
 
-  for (unsigned int i=0; i<_variable_names.size(); i++)
-    _values_tmp[i].push_back(values[i]);
+  mooseAssert(values.size() == _variable_names.size(), "Mismatch of variable names to vector size");
+  for (auto i = beginIndex(values); i < values.size(); ++i)
+    _values[i]->push_back(values[i]);
 }
 
 void
 SamplerBase::initialize()
 {
-  _x_tmp.clear();
-  _y_tmp.clear();
-  _z_tmp.clear();
-  _id_tmp.clear();
-
-  for (unsigned int i=0; i<_variable_names.size(); i++)
-    _values_tmp[i].clear();
+  _x.clear();
+  _y.clear();
+  _z.clear();
+  _id.clear();
+
+  std::for_each(_values.begin(), _values.end(),
+                [](VectorPostprocessorValue * vec)
+                {
+                  vec->clear();
+                });
 }
 
 void
 SamplerBase::finalize()
 {
-  // Get the values from everywhere
-  _comm.allgather(_x_tmp, false);
-  _comm.allgather(_y_tmp, false);
-  _comm.allgather(_z_tmp, false);
-  _comm.allgather(_id_tmp, false);
-
-  for (unsigned int i=0; i<_variable_names.size(); i++)
-    _comm.allgather(_values_tmp[i], false);
-
-  // Next... figure out the correct sorted positions of each value
-  std::vector<size_t> sorted_indices;
-
-  switch (_sort_by)
-  {
-    case 0: // x
-      Moose::indirectSort(_x_tmp.begin(), _x_tmp.end(), sorted_indices);
-      break;
-    case 1: // y
-      Moose::indirectSort(_y_tmp.begin(), _y_tmp.end(), sorted_indices);
-      break;
-    case 2: // z
-      Moose::indirectSort(_z_tmp.begin(), _z_tmp.end(), sorted_indices);
-      break;
-    case 3: // id
-      Moose::indirectSort(_id_tmp.begin(), _id_tmp.end(), sorted_indices);
-      break;
-  }
-
-  // Use the sorted_indices to copy out of the temporary vectors and into the actual output vectors
-  for (unsigned int i=0; i<_variable_names.size(); i++)
+  /**
+   * We have several vectors that all need to be processed in the same way.
+   * Rather than enumerate them all, let's just create a vector a pointers
+   * and work on them that way.
+   */
+  // Initialize the pointer vector with the position and ID vectors since we sort only on these.
+  std::vector<VectorPostprocessorValue *> vec_ptrs { { &_x, &_y, &_z, &_id } };
+  std::vector<std::size_t> sorted_indices;
+
+  // Gather up each of the partial vectors then choose one to sort by
+  for (auto i = beginIndex(vec_ptrs); i < vec_ptrs.size(); ++i)
   {
-    _values[i]->resize(sorted_indices.size());
+    _comm.allgather(*vec_ptrs[i], false);
 
-    for (unsigned int j=0; j<sorted_indices.size(); j++)
-      (*_values[i])[j] = _values_tmp[i][sorted_indices[j]];
+    // Sort by the vector chosen by the user
+    if (i == _sort_by)
```

Comment:
wow - missed that. This will allow me to clean up the code quite a bit more (I see further collapsing cascades after this.